### ISSUE TITLE
Context("Point") accepts [a,b] and {a,b} as correct.

### ISSUE THREAD

**Issue by somiaj** [openwebwork/pg#535]
When using Context("Point"), all answers (a,b), [a,b], and {a,b} are accepted as correct. In some cases (working with sets vs order pairs in my case), I found out students were entering in {a,b} as elements of a Cartesian product and being marked correct. I have written a custom grader that rejects answers that have [] or {} in their string, but think there should at least be an option (maybe there is but I didn't see it) to reject points written with [] or {}.

Here is an example problem, with my custom grader included (unsure if there is a better way to achieve what I did, and I had trouble writing a single regex line that worked, that is why I split it into multiple conditions).

```
DOCUMENT();      

loadMacros(
   "PGstandard.pl",
   "MathObjects.pl",
   "PGcourse.pl"
);

TEXT(beginproblem());

Context("Point");
$ans1 = Compute("(5,7)");
$ans2 = Compute("(3,-2)");
$cmp2 = sub {
  my ( $c, $s, $ansHash ) = @_;

  $sstr = $s->string;
  if ( ($sstr =~ /[\[]/)||($sstr =~ /[\]]/)||($sstr =~ /{/)||($sstr =~/}/) )
  { return 0; }
  if ( $c == $s ) { return 1; }
  else { return 0; }

};

Context()->texStrings;
BEGIN_TEXT

Enter in the point \( (5,7) \). Answer will accept
\( \lbrace 5,7 \rbrace \) and \( [5,7] \).
$BR
\{ ans_rule(5) \}

$PAR
Enter in the point \( (3,-2) \). Answer will not accept
\( \lbrace 3,-2 \rbrace \) or \( [3,-2] \).
$BR
\{ ans_rule(5) \}

END_TEXT
Context()->normalStrings;

ANS($ans1->cmp);
ANS($ans2->cmp(checker => $cmp2));

ENDDOCUMENT();
```

**Comment by Alex-Jordan** [comment]
Hello @somiaj  This is a user support question, which will get good responses if you post it in the WeBWorK forums:
https://webwork.maa.org/moodle/mod/forum/index.php?id=3

You may need to sign up for that if you don't already have an account. 

Since I have a quick answer for you, I'll mention it here. You can change the way the parser views delimiters following this:
https://webwork.maa.org/wiki/Modifying_Contexts_(advanced)#Lists_and_Delimiters

**Comment by drgrice1** [comment]
@Alex-Jordan:  I think that this is a valid issue to file here.  I agree that the Point context should not accept all of those answers.  I have encountered this before, and have implemented what I would consider hacks to avoid this.

@somiaj:  The hacks I implemented are something like this:
```
Context("Point");
Context()->parens->set(
    '{' => {close => '}', type => 'List', removable => 1},
    '[' => {close => ']', type => 'List', removable => 1}
);
```
You could instead use type `'Matrix'` for the brackets.

**Comment by Alex-Jordan** [comment]
Hmm, if someone's position is that it was a mistake for the Point context to use those delimiters as indicators of a Point object, then I can understand that viewpoint as a undesirable behavior from the `pg` repo.

But I don't think I agree with that viewpoint. There can be classroom contexts (I don't mean in the PG sense) where the custom is to use these as point delimiters. And more importantly, if it's the "Point" context, you are saying something about the expectation for what kind of answers will come. If it's a context that is suppose to handle set answers and point answers together, maybe that is not what "Point" was meant for.

Either @dpvc was thinking along those lines when setting up the Point context defaults, or it was an oversight. I would defer to what Davide thinks about it.

But I concede that it's not as clear cut of a support issue as I thought at first.

**Comment by drgrice1** [comment]
Although there are contexts where someone may use bracket or brace delimiters for a point object, that is certainly a non-standard convention.  I would think that it would be better to use standard conventions for the defaults, and when someone wants non-standard behavior, then they should be the ones that implement the changes.

**Comment by somiaj** [comment]
Hello,

Thanks for the quick response, and to me it is an issue that set brackets are accepted as delimiters for a point by default. Even on the link you shared to delimiters, it is not clear that {} can be used with points. The table provided there clearly shows () for points and {} for sets.

On the other hand, for sets you have to manually add other delimiters. A Set as part of the Interval context will not accept (a,b) for the set {a,b} (though this might be a by-product that (a,b) is an interval and not an ordered pair).

I also think one should have to add nonstandard delimiters (vs remove extra ones) if they are in a situation where they want to allow {a,b} or [a,b] as points.

**Comment by Alex-Jordan** [comment]
Here is where the Point context defaults are set:
https://github.com/openwebwork/pg/blob/master/lib/Parser/Context/Default.pm#L307

It copies the Vector context and tweaks some things. Vector in turn is copied from Full. And here:
https://github.com/openwebwork/pg/blob/master/lib/Parser/Context/Default.pm#L109

is where brackets and braces are ultimately being set to make a Point by default.

**Comment by dpvc** [comment]
It is hard to remember exactly my thinking from 15 years ago (which was before many of the MathObject use cases were developed), but I think the idea was that, in the Numeric context, parentheses, brackets, and braces all work like parentheses.  So you can do `2 - [x*(1+x)]` or `{x+5}/[1-x]`.  So these three all should be treated the same, and since parentheses produce points, so should the others.

That may have been a bad decision, and it may make sense to change brackets and braces (and parentheses, too) to produce lists in Numeric context.  Then change parentheses to produce points in Vector contexts.  That might be more intuitive, especially in situations that involve sets.

**Comment by somiaj** [comment]
Just wanting to fix @dgrice1 hack to remove the {} and [] delimiters, missing the option `formList => 0`, in case anyone else finds this looking for solution until the defaults are changed.

```
Context()->parens->set(
    '{' => {close => '}', type => 'List', removable => 1, formList => 0},
    '[' => {close => ']', type => 'List', removable => 1, formList => 0}
);
```

**Comment by drgrice1** [comment]
Why do you need to add `formList => 0`?  That is the default.

**Comment by somiaj** [comment]
Because my tests seemed to fail without it, but I changed something else to make my tests easier to do (changed from a List of points to a single point), which changed things. Your fix only affects single points, not lists of points. So I confused myself thinking it didn't work, then simplified my tests, not realizing that changed things. I'll delete my change.

**Comment by somiaj** [comment]
@dgrice1 I have done some more testing, and unsure if this is another bug/issue or not. But with using your 'hack'

```
Context("Point");
Context()->parens->set(
    '{' => {close => '}', type => 'List', removable => 1},
    '[' => {close => ']', type => 'List', removable => 1}
);

$ans1 = Point("(3,5)");
$ans2 = List("(3,5),(4,6)");
```

`$ans1->cmp` will correctly reject points of the form {a,b}, but `$ans2->cmp` will not (hence why I had issues with your fix initially and thought it wasn't working). But if I write a custom list checker according to <https://webwork.maa.org/wiki/Custom_Answer_Checkers_for_Lists>, the check `if ($p->type ne "Point")` will correctly reject points of the form {a,b}.

So could this be a bug in the default List checker that isn't correctly identifying the types of objects in the list when checking them? At least in this case the default List checker isn't honoring the Context()->parens->set change.

Thanks again for your support.

**Comment by drgrice1** [comment]
First, I guess I should say that the `Context()->parens->set(...)` thing is not necessarily a hack.  It depends on what is intended for the defaults of the `Point` context.

As to whether that is a bug or not, I will defer to @dpvc.

**Comment by dpvc** [comment]
I would say it is probably a bug.  The issue is that the list checker uses the correct answer's `cmp_compare()` function to determine if it equals one of the student answers, but `cmp_compare()` doesn't perform the `typeMatch()` that normally would rule out a point matching a list, and the list comparison promotes the point to a list automatically.

This line

https://github.com/openwebwork/pg/blob/a284c3b31a26fb316c2fa6b34246df8ba2895a3e/lib/Value/AnswerChecker.pm#L1529

probably should be

``` perl
	if ($correct[$k]->typeMatch($entry) && $correct[$k]->cmp_compare($entry,$ans,$nth,$value)) {
```

and a similar change to 

https://github.com/openwebwork/pg/blob/a284c3b31a26fb316c2fa6b34246df8ba2895a3e/lib/Value/AnswerChecker.pm#L1520

This will prevent the comparison from being made if the student answer doesn't have a type that is compatible with the correct answer.  This **is** a significant change to the list answer checking, and it may affect how some problems are checked, since people may have been relying on this (incorrect) behavior in some way (I can't think of one, but since this has been the way it has worked for 15 years, someone undoubtedly relies on it).  For example, when we made a similar change recently to have Union objects not accept lists of intervals instead of explicit unions of them, we found out that some people were taking advantage of that error to give partial credit.  I'm not sure the best way to check if any problems do something similar for general lists.

@somiaj, note that you don't need to replace the list checker itself, but only the entry checker (`checker =>` rather than `list-checker =>' as this is an entry-by-entry issue, not a list-as-a-whole issue.  That should make for a simpler change.

Also, note that you only need to do

``` perl
Context()->parens->set(
    '{' => {type => 'List'},
    '[' => {type => 'List'}
);
```

as the other values are already the defaults (the `set` method only **adds** to the definition, it doesn't replace it).

**Comment by drgrice1** [comment]
> For example, when we made a similar change recently to have Union objects not accept lists of intervals instead of explicit unions of them, we found out that some people were taking advantage of that error to give partial credit. I'm not sure the best way to check if any problems do something similar for general lists.

By "some people" @dpvc means me.

It is likely I have problems that are relying on this bug also.  I can work with it though if the bug is fixed.  I am sure I can find a work around as I did for the Union objects.

**Comment by dpvc** [comment]
@drgrice1, one approach would be to subclass the correct answer class and override the `typeMatch()` method to allow comparisons with any type, which would effectively make the changed line the same as it was originally.

**Comment by dpvc** [comment]
I made a pull request that includes the changes we have been talking about, in case you want to test that out.

**Comment by somiaj** [comment]
Thanks, I don't have a development server set up at the moment (a todo list) and don't want to make changes to my server until after the semester, so can't provide you any tests at the moment. I do have fixes that are working for my use case, and will clean them up once the semester is over. Thanks again for the support.

**Comment by pstaabp** [comment]
Now that we are 10 months out from this and have merged in @dpvc PR #535.  Do we want to close this or keep the discussion open?
